### ISSUE TITLE
Disantangle long term memory update and reference frame

### ISSUE THREAD

**Issue by bhack** [yoxu515/aot-benchmark#37]
Can we disentangle long term memory update and reference frame?

https://github.com/yoxu515/aot-benchmark/blob/f9a62b60e12218019c0617efa1315dcc73147cdb/networks/engines/aot_engine.py#L243-L246

What I think we could add an extra parameter to the function signature for updating or not the long term memory.

Cause currently it is hard to make multiple evaluation passes with different semi-supervision as the same function it is required also for preparing `curr_enc_embs`.

**Comment by bhack** [comment]
I was trying to add all the reference keyframes at the beginning it is why it could be useful to disentangle the long term memory update step.
Do you see any side effect in the long term memory logic adding all the reference keyframes at the beginning?

**Comment by z-x-yang** [comment]
Your implementation of such a method is feasible. Generally speaking, the more reference frames there are in long-term memory, the better the overall performance will be.

**Comment by z-x-yang** [comment]
I plan to close this issue. But if you still have some problems, please reopen it.

**Comment by bhack** [comment]
It would be nice if you could to add this flexibility to that function so that we are in control of the number of reference frame and experiment with different level of supervision.

**Comment by bhack** [comment]
I think that it is partially connected to what another user requested at https://github.com/yoxu515/aot-benchmark/issues/26

Is there any difference adding at the engine "warm-up" or just in the specific frame in the sequence?

**Comment by z-x-yang** [comment]
You may add additional reference frames by invoking the add_reference_frame function multiple times, ultimately appending the first frame through the same method. The purpose of add_reference_frame is to incorporate the input frame into the long-term memory and substitute the short-term memory.

**Comment by z-x-yang** [comment]
In YouTube-VOS, many sequences have multiple reference frames.

**Comment by bhack** [comment]
The problem is that you are going to double the memory requirements if you call multiple time `add_reference_frame` at frame 0 and at runtime on the sequence, as `add_reference_frame` always call   `self.update_long_term_memory(lstt_long_memories)`.

**Comment by z-x-yang** [comment]
Why not call at frame 0 for only once? Sorry, I don't get you.在 2023年4月4日，04:29，bhack ***@***.***> 写道：﻿
The problem is that you are going to double the memory requirements if you call multiple time add_reference_frame at frame 0 and at runtime on the sequence, as add_reference_frame always call   self.update_long_term_memory(lstt_long_memories).

**Comment by bhack** [comment]
Suppose that on a sequence of 100 frames at frame 0 you add 10 reference frames where the  last reference is 90.
What you will do then in the sequence loop?

Are you going to only  `match_propogate_one_frame` at frame 1,2,3 ...100?

**Comment by z-x-yang** [comment]
I have added a new control variable ([skip_long_term_update](https://github.com/yoxu515/aot-benchmark/blob/8d6192c679b4b6885aaa664a3475e92a437024fd/networks/engines/aot_engine.py#L625)) to allow skipping the update of long-term memory during a single update. You may utilize this to skip updating frames that have already been added to the reference frames, thus avoiding redundant memory updates.

**Comment by bhack** [comment]
Thanks. it is really to what I've tried locally. 
Is there any other side effect about the order of the reference keyframes we have in the long term memory? 
E.g. at frame 0 to predict 1 in the runtime loop you have in the engine frame 90 as the last added keyframe in the long term memory.

**Comment by bhack** [comment]
What I meant has this an impact over the number of attention heads?

**Comment by z-x-yang** [comment]
Single-head attention leads to better efficiency but loses accuracy. BTW, head number is not adaptable during inference, corresponding training is necessary.在 2023年4月11日，17:21，bhack ***@***.***> 写道：﻿
What I meant has this an impact over the number of attention heads?

**Comment by bhack** [comment]
> BTW, head number is not adaptable during inference, corresponding training is necessary.

Yes I understand what I mean, just to make an example, that we have a bank of 5 reference frames with your last commit, over a sequence of 100. We have 0, 25, 50, 100 reference in the long term memory. 
When we are going to predict frame 1 or frame 26 do we have any other side effect about accumulating all these reference keyframe in the "warm-up" phase? Is there a limit on the long term memory frame we could really accumulate other then the memory limit?
In the paper:
> Concretely, the computational complexity of long-term attention is O(N T H2W 2), which is proportional to the head number N since each
head contains a correlation function, Corr(Q, K).

What is the role of N where we have already accumulated T frames at the engine bootstrap?

**Comment by z-x-yang** [comment]
For each head, there is an individual calculation of SoftMax in Core(Q, K), which is a non-linear function.

**Comment by bhack** [comment]
Do you meant that the order on which we add reference frames in the engine warmup will not going to impact the accuracy at runtime with the new `skip_long_term_update=True`?

**Comment by z-x-yang** [comment]
The frame order in the long-term memory should not affect the performance.
